### DISCUSSION TITLE
Language runtime RAM vs TIC-80 RAM

### DISCUSSION THREAD

**Discussion by Liorst4** [nesbox/TIC-80#2092] [Q&A] [answered]
Why use the dedicated ram of TIC-80, when runtimes provide an huge heap to work with?

In the manual its stated that there is 96KB of RAM. (https://tic80.com/learn)
Address 14E24 and up are free for the user (~10KB).

On the other hand, the JS runtime can create a way bigger heap.

https://github.com/nesbox/TIC-80/blob/ca3bb11f7d91b6f61c30b27e95958521e3432576/src/api/js.c#L1042

https://duktape.org/api.html#duk_create_heap

Since `alloc_func`, `realloc_func` and `free_func` are `NULL`, it ends up using the libc heap (`malloc`, `realloc`, `free`), which as far as I know is unbound by the 96KB limit.

Nothing is stopping the user from creating a 1000000 item array.
What uses are there for `peek` and `poke` on 14E24 and up?

**Comment by joshgoebel** [comment, marked as answer]
> Address 14E24 and up are free for the user (~10KB).

_Please point me to where it still says this._  This is no longer true, all that memory is declared RESERVED now for future use by the hardware (see Wiki).  One should indeed primary use runtime memory and variables.  The only reason to use real RAM would be some type of "high performance" hardware tricks you wanted to do by perhaps coming RAM to and from the VRAM in large chunks with `memcpy` which the API can do natively orders of magnitude faster than a runtime.

**Reply by Liorst4** [reply]
https://tic80.com/learn

```
+-----------------------------------+
|           96KB RAM LAYOUT         |
+-------+-------------------+-------+
| ADDR  | INFO              | BYTES |
+-------+-------------------+-------+
| 00000 | <VRAM>            | 16384 |
| 04000 | TILES             | 8192  |
| 06000 | SPRITES           | 8192  |
| 08000 | MAP               | 32640 |
| 0FF80 | GAMEPADS          | 4     |
| 0FF84 | MOUSE             | 4     |
| 0FF88 | KEYBOARD          | 4     |
| 0FF8C | SFX STATE         | 16    |
| 0FF9C | SOUND REGISTERS   | 72    |
| 0FFE4 | WAVEFORMS         | 256   |
| 100E4 | SFX               | 4224  |
| 11164 | MUSIC PATTERNS    | 11520 |
| 13E64 | MUSIC TRACKS      | 408   |
| 13FFC | MUSIC STATE       | 4     |
| 14000 | STEREO VOLUME     | 4     |
| 14004 | PERSISTENT MEMORY | 1024  |
| 14404 | SPRITE FLAGS      | 512   |
| 14604 | FONT              | 1016  |
| 149FC | FONT PARAMS       | 8     |
| 14A04 | ALT FONT          | 1016  |
| 14DFC | ALT FONT PARAMS   | 8     |
| 14E04 | BUTTONS MAPPING   | 32    |
| 14E24 | ... (free)        | 12764 |                    <--------------------------------------- THIS
+-------+-------------------+-------+
```

**Reply by Liorst4** [reply]
Also in the console, when running `help ram` 

https://github.com/nesbox/TIC-80/blob/ca3bb11f7d91b6f61c30b27e95958521e3432576/src/studio/screens/console.c#L3031

**Comment by joshgoebel** [comment]
I can make a PR for `help ram`, but we need @nesbox to get the website updated...

**Reply by nesbox** [reply]
I'll update the website after the PR.

**Comment by joshgoebel** [comment]
See #2094
